### PR TITLE
Fw update improvements

### PR DESCRIPTION
**Pull request by macpijan**

### CODE REVIEW

File: meta-dts-distro/recipes-dts/dts/dasharo-deploy/dasharo-deploy [hunk 1]
```diff
@@ -104,7 +30,6 @@ PROGRAMMER_EC="ite_ec"
 [ -z "$SYSTEM_MODEL" ] && error_exit "SYSTEM_MODEL not given"
 
 CMD="$1"
-CMD_CLOUD_LIST="/usr/sbin/cloud_list"
 board_config() {
```
Comment (TomaszAIR) [inline meta-dts-distro/recipes-dts/dts/dasharo-deploy/dasharo-deploy:33]: @macpijan looks like we are using the same binaries for ec_transition, could this function also go to dts-functions.sh as a common part? Or maybe we can get rid off ec_transition script at all? What is the logical difference between EC transition and initial deployment of Dasharo on platforms like NV4x and NS50? Is there a way to change EC only without touching BIOS? If someone wants to use Dasharo on these laptops, can he do it without changing EC?

Comment (TomaszAIR) [reply meta-dts-distro/recipes-dts/dts/dasharo-deploy/dasharo-deploy:33]: There are only slight difference in board config for NV4x

```
PROGRAMMER_BIOS="internal:boardmismatch=force"
PROGRAMMER_EC="ite_ec:boardmismatch=force,romsize=128K,autoload=disable"
```

**Review by TomaszAIR — CHANGES_REQUESTED** [review]
I left one or two comments, please change base to `develop` and rebase

File: meta-dts-distro/recipes-dts/dts/dts/dts-environment.sh [hunk 2]
```diff
@@ -6,6 +6,7 @@ BOARD_MODEL="$(dmidecode -s baseboard-product-name)"
 
 BIOS_VENDOR="$(dmidecode -s bios-vendor)"
 BIOS_VERSION="$(dmidecode -s bios-version)"
+DASHARO_VERSION="$(echo $BIOS_VERSION | cut -d ' ' -f 3 | tr -d 'v')"
```
Comment (TomaszAIR) [inline meta-dts-distro/recipes-dts/dts/dts/dts-environment.sh:9]: For Dasharo fw used on optiplex
```
# dmidecode -s bios-version
Dasharo (coreboot+SeaBIOS)
```

So DASHARO_VERSION is ""

File: meta-dts-distro/recipes-dts/dts/dasharo-deploy/dasharo-deploy [hunk 3]
```diff
@@ -617,6 +574,128 @@ install() {
...
+  error_check "Failed to update EC firmware"
+}
+
+compare_versions() {
+    # return 1 if ver2 > ver1
+    # return 0 otherwise
+
+    local ver1="$1"
+    local ver2="$2"
+
```
Comment (TomaszAIR) [inline meta-dts-distro/recipes-dts/dts/dasharo-deploy/dasharo-deploy:596]: @macpijan can we check here if ver1 and ver2 match regex?

`if [[ $ver1 =~ ^[0-9]+\.[0-9]+\.[0-9]+$ ]] && [[ $ver2 =~ ^[0-9]+\.[0-9]+\.[0-9]+$ ]]; then`

that should be ok?

File: meta-dts-distro/recipes-dts/dts/dasharo-deploy/dasharo-deploy [hunk 4]
```diff
@@ -617,6 +574,128 @@ install() {
...
+  echo "Downloading Dasharo firmware update..."
+  curl --progress-bar -f "$BIOS_LINK" -o $BIOS_UPDATE_FILE
+  error_check "Cannot access $FW_STORE_URL. Please check your internet connection"
+  echo "Checking sha256sum..."
+  sha256sum --check <(echo "$BIOS_HASH")
+  error_check "Failed to verify Dasharo firmware"
+
+  if [ "$HAVE_EC" = "true" ]; then
+    system76_ectool info >> $ERR_LOG_FILE 2>&1
+    if [ $? -ne 0 ]; then
```
Comment (TomaszAIR) [inline meta-dts-distro/recipes-dts/dts/dasharo-deploy/dasharo-deploy:656]: Can we use error_check here?

File: meta-dts-distro/recipes-dts/dts/dasharo-deploy/dasharo-deploy [hunk 5]
```diff
@@ -617,6 +574,128 @@ install() {
...
+  sleep 1
+  echo "3..."
+  sleep 1
+  echo "2..."
+  sleep 1
+  echo "1..."
+  sleep 0.5
+  echo "Rebooting"
+  sleep 1
+  ${CMD_REBOOT}
```
Comment (TomaszAIR) [inline meta-dts-distro/recipes-dts/dts/dasharo-deploy/dasharo-deploy:696]: @macpijan reboot is enough even if we flash ec?

File: meta-dts-distro/recipes-dts/dts/dts/dts-functions.sh [hunk 6]
```diff
@@ -1,9 +1,121 @@
...
+      echo 'No network connection to 3mdeb cloud, please recheck Ethernet connection'
+      return 1
+    fi
+    sleep 1
+  done
+  return 0
+}
+
+check_se_creds() {
+  # TODO: what is the difference between tpmDownloadURL and tmpDOWNLOADURL ?!
```
Comment (TomaszAIR) [inline meta-dts-distro/recipes-dts/dts/dts/dts-functions.sh:103]: Looks like I overlooked this with the previous review

File: meta-dts-distro/recipes-dts/dts/dasharo-deploy/dasharo-deploy (hunk 1 above)
Comment (macpijan) [reply meta-dts-distro/recipes-dts/dts/dasharo-deploy/dasharo-deploy:37]: > What is the logical difference between EC transition and initial deployment of Dasharo on platforms like NV4x and NS50? 

Initial deployment deploys both BIOS and EC Dasharo firmware on machine which was previously using stock firmwares (both BIOS and EC).

Transition was meant to update Dasharo BIOS and **transition** EC firwmare fro proprietary to open Dasharo EC firmware.

> Is there a way to change EC only without touching BIOS

It is, but not recommended. We should not give such option for end-users. For end users we should always keep both firmwares in sync.

> If someone wants to use Dasharo on these laptops, can he do it without changing EC?

Not really. He could install <1.3.0 releases, but we do not recommend it, of course.

File: meta-dts-distro/recipes-dts/dts/dasharo-deploy/dasharo-deploy [hunk 7]
```diff
@@ -92,54 +20,57 @@ FW_BACKUP_TAR="${FW_BACKUP_DIR}.tar.gz"
...
 CMD="$1"
-CMD_CLOUD_LIST="/usr/sbin/cloud_list"
 board_config() {
   case "$BOARD_VENDOR" in
     "Notebook")
       case "$SYSTEM_MODEL" in
         "NV4XMB,ME,MZ")
-          DASHARO_REL_NAME="novacustom_nv4x"
-          DASHARO_REL_VER="1.3.0"
+          DASHARO_REL_NAME="novacustom_nv4x_tgl"
```
[the remark this answers is not in the document]
Comment (macpijan) [reply meta-dts-distro/recipes-dts/dts/dasharo-deploy/dasharo-deploy:42]: @TomaszAIR Definitely

File: meta-dts-distro/recipes-dts/dts/dts/dts-environment.sh (hunk 2 above)
Comment (macpijan) [reply meta-dts-distro/recipes-dts/dts/dts/dts-environment.sh:9]: I guess we need to fix optiplex

File: meta-dts-distro/recipes-dts/dts/dasharo-deploy/dasharo-deploy (hunk 5 above)
Comment (macpijan) [reply meta-dts-distro/recipes-dts/dts/dasharo-deploy/dasharo-deploy:733]: We will never reach this code when we update EC. system76_tool flash automatically shuts down after flashing was done.

File: meta-dts-distro/recipes-dts/dts/dts/dts-environment.sh (hunk 2 above)
Comment (TomaszAIR) [reply meta-dts-distro/recipes-dts/dts/dts/dts-environment.sh:9]: for sure

File: meta-dts-distro/recipes-dts/dts/dasharo-deploy/dasharo-deploy (hunk 7 above)
[the remark this answers is not in the document]
Comment (TomaszAIR) [reply meta-dts-distro/recipes-dts/dts/dasharo-deploy/dasharo-deploy:42]: https://github.com/Dasharo/meta-dts/pull/29/files#diff-ac1fdd7af29e82bf8d7ca23462952d638564fce3189e36cbb6eebeb48f480d20R76 added, should be also tested

File: meta-dts-distro/recipes-dts/dts/dasharo-deploy/dasharo-deploy (hunk 3 above)
Comment (TomaszAIR) [reply meta-dts-distro/recipes-dts/dts/dasharo-deploy/dasharo-deploy:629]: added https://github.com/Dasharo/meta-dts/pull/29/files?diff=split&w=0#diff-ac1fdd7af29e82bf8d7ca23462952d638564fce3189e36cbb6eebeb48f480d20R630

File: meta-dts-distro/recipes-dts/dts/dasharo-deploy/dasharo-deploy (hunk 4 above)
Comment (TomaszAIR) [reply meta-dts-distro/recipes-dts/dts/dasharo-deploy/dasharo-deploy]: fixed https://github.com/Dasharo/meta-dts/pull/29/commits/4d7244392eb909e4f4fa4709d32547c20cfce84a

File: meta-dts-distro/recipes-dts/dts/dts/dts-environment.sh (hunk 2 above)
Comment (TomaszAIR) [reply meta-dts-distro/recipes-dts/dts/dts/dts-environment.sh:9]: version syntax added https://github.com/Dasharo/meta-dts/pull/29/commits/29fd7b1f4d992c080e398732a27ed147a6fc0953#diff-ac1fdd7af29e82bf8d7ca23462952d638564fce3189e36cbb6eebeb48f480d20R630 so in such cases we should know if something bad happened

File: meta-dts-distro/recipes-dts/dts/dts/dts-functions.sh (hunk 6 above)
Comment (TomaszAIR) [reply meta-dts-distro/recipes-dts/dts/dts/dts-functions.sh:146]: This TODO will be handled in next release

File: meta-dts-distro/recipes-dts/dts/dasharo-deploy/dasharo-deploy (hunk 1 above)
Comment (TomaszAIR) [reply meta-dts-distro/recipes-dts/dts/dasharo-deploy/dasharo-deploy:37]: ec_transition was moved into Dasharo update https://github.com/Dasharo/meta-dts/pull/29/commits/37b0a3f010b63361e06ce1ecb64f710a7bda33f9

File: meta-dts-distro/recipes-dts/dts/dasharo-deploy/dasharo-deploy [hunk 8]
```diff
@@ -424,99 +381,99 @@ backup() {
...
 
 resign_binary()
 {
   if [ "$HAVE_VBOOT" -eq 0 ]; then
     download_keys
-    sign_firmware.sh $BIOS_UPDATE_DIR $KEYS_DIR $RESIGNED_BIOS_UPDATE_DIR
+    sign_firmware.sh $BIOS_UPDATE_FILE $KEYS_DIR $RESIGNED_BIOS_UPDATE_DIR
     error_check "Cannot resign binary file. Please, make sure if you have proper keys. Aborting..."
-    BIOS_UPDATE_DIR="$RESIGNED_BIOS_UPDATE_DIR"
+    BIOS_UPDATE_FILE="$RESIGNED_BIOS_UPDATE_DIR"
```
[the remark this answers is not in the document]
Comment (TomaszAIR) [reply meta-dts-distro/recipes-dts/dts/dasharo-deploy/dasharo-deploy]: fixed https://github.com/Dasharo/meta-dts/pull/29/files#diff-ac1fdd7af29e82bf8d7ca23462952d638564fce3189e36cbb6eebeb48f480d20R482

File: meta-dts-distro/recipes-dts/dts/dasharo-deploy/dasharo-deploy [hunk 9]
```diff
@@ -424,99 +381,99 @@ backup() {
...
+    cbfstool $BIOS_UPDATE_FILE truncate -r FW_MAIN_B
   fi
 }
 
 resign_binary()
 {
   if [ "$HAVE_VBOOT" -eq 0 ]; then
     download_keys
-    sign_firmware.sh $BIOS_UPDATE_DIR $KEYS_DIR $RESIGNED_BIOS_UPDATE_DIR
+    sign_firmware.sh $BIOS_UPDATE_FILE $KEYS_DIR $RESIGNED_BIOS_UPDATE_DIR
```
[the remark this answers is not in the document]
Comment (TomaszAIR) [reply meta-dts-distro/recipes-dts/dts/dasharo-deploy/dasharo-deploy]: fixed https://github.com/Dasharo/meta-dts/pull/29/files#diff-ac1fdd7af29e82bf8d7ca23462952d638564fce3189e36cbb6eebeb48f480d20R480

File: meta-dts-distro/recipes-dts/dts/dasharo-deploy/dasharo-deploy [hunk 10]
```diff
@@ -583,7 +540,7 @@ install() {
   fi
 
   echo "Installing Dasharo firmware..."
-  flashrom -p "$PROGRAMMER_BIOS" ${FLASH_CHIP_SELECT} ${FLASHROM_ADD_OPT} -w "$BIOS_UPDATE_DIR"
+  flashrom -p "$PROGRAMMER_BIOS" ${FLASH_CHIP_SELECT} ${FLASHROM_ADD_OPT_DEPLOY} -w "$BIOS_UPDATE_DIR"
```
[the remark this answers is not in the document]
Comment (TomaszAIR) [reply meta-dts-distro/recipes-dts/dts/dasharo-deploy/dasharo-deploy]: fixed https://github.com/Dasharo/meta-dts/pull/29/files#diff-ac1fdd7af29e82bf8d7ca23462952d638564fce3189e36cbb6eebeb48f480d20R586

File: meta-dts-distro/recipes-dts/dts/dasharo-deploy/dasharo-deploy [hunk 11]
```diff
@@ -152,29 +83,50 @@ board_config() {
...
               NEED_EC_RESET="false"
-              BIOS_HASH_PROD="b7a0ca05a860c27c8357ed29c51317f1c2ace0590f073f65978296d2036e68dd  $BIOS_UPDATE_DIR"
+              BIOS_HASH_PROD="$(wget -O /dev/stdout ${BIOS_LINK_PROD}.sha256 2>>$ERR_LOG_FILE | cut -d ' ' -f 1) $BIOS_UPDATE_FILE"
               NEED_SMBIOS_MIGRATION="true"
               NEED_BLOB_TRANSMISSION="false"
               PROGRAMMER_BIOS="internal"
               PROGRAMMER_EC=""
-              FLASHROM_ADD_OPT="--ifd -i bios"
+              FLASHROM_ADD_OPT_DEPLOY="--ifd -i bios"
+              FLASHROM_ADD_OPT_UPDATE="--fmap -i RW_SECTION_A -i RW_SECTION_B"
```
[the remark this answers is not in the document]
Comment (TomaszAIR) [reply meta-dts-distro/recipes-dts/dts/dasharo-deploy/dasharo-deploy]: fixed https://github.com/Dasharo/meta-dts/pull/29/files#diff-ac1fdd7af29e82bf8d7ca23462952d638564fce3189e36cbb6eebeb48f480d20R130

**Review by TomaszAIR — APPROVED** [review]
Approved and tested

Initial deployment and update on MSI Z690-A DDR4
update on  MSI Z690-A DDR5, NV4X TGL, NS5X TGL

**Pull request merged**